### ISSUE TITLE
E1019 error incorrectly reported for Fn::Sub function in AWS::CertificateManager::Certificate resource on cfn-lint v1.5.2

### ISSUE THREAD

**Issue by JHBoricua** [aws-cloudformation/cfn-lint#3485]
### CloudFormation Lint Version

1.5.2

### What operating system are you using?

Amazon Linux 2

### Describe the bug

I'm encountering an issue with cfn-lint, where it's reporting an E1019 error for the Fn::Sub function in an AWS::CertificateManager::Certificate resource, even though the function is receiving an array as expected.

Here's the relevant part of my CloudFormation template:

```
  Certificate:
    Condition: AccountIsCorrect
    Type: AWS::CertificateManager::Certificate
    Properties:
      CertificateAuthorityArn: !Sub 'arn:${AWS::Partition}:acm-pca:us-east-1:myCaARN'
      DomainName: !Sub
        - '${CertificateCN}-${SdlcKey}.example.com'
        - CertificateCN: !FindInMap [ Global, ALB, CertificateCN ]
      SubjectAlternativeNames: !If
        - IsProd
        - - !Sub
            - '${CertificateCN}-${SdlcEnvironment}.example.com'
            - CertificateCN: !FindInMap [ Global, ALB, CertificateCN ]
              SdlcEnvironment: !FindInMap [ SdlcKey, !Ref SdlcKey, SdlcEnvironment ]
          - !Sub
            - '${CertificateCN}.example.com'
            - CertificateCN: !FindInMap [ Global, ALB, CertificateCN ]
        - !Sub
          - '${CertificateCN}-${SdlcEnvironment}.example.com'
          - CertificateCN: !FindInMap [ Global, ALB, CertificateCN ]
            SdlcEnvironment: !FindInMap [ SdlcKey, !Ref SdlcKey, SdlcEnvironment ]
```

The error message I'm seeing is:

```
[cfn-lint] E1019: {'Fn::Sub': ['${CertificateCN}-${SdlcEnvironment}.example.com', {'CertificateCN': {'Fn::FindInMap': ['Global', 'ALB', 'CertificateCN']}, 'SdlcEnvironment': {'Fn::FindInMap': ['SdlcKey', {'Ref': 'SdlcKey'}, 'SdlcEnvironment']}}]} is not of type 'array'
```

I believe this is a false positive, as the Fn::Sub function is indeed receiving an array. I'm using cfn-lint version 1.5.2. I cannot reproduce this when running against cfn-lint v0.87.9.

### Expected behavior

cfn-lint should correctly recognize the Fn::SUB function as receiving an array, and not report an E1019 error for this usage of the function. Instead, it should validate the template without any errors, assuming all other parts of the template are correct.

### Reproduction template

```
  Certificate:
    Condition: AccountIsCorrect
    Type: AWS::CertificateManager::Certificate
    Properties:
      CertificateAuthorityArn: !Sub 'arn:${AWS::Partition}:acm-pca:us-east-1:myCaARN'
      DomainName: !Sub
        - '${CertificateCN}-${SdlcKey}.example.com'
        - CertificateCN: !FindInMap [ Global, ALB, CertificateCN ]
      SubjectAlternativeNames: !If
        - IsProd
        - - !Sub
            - '${CertificateCN}-${SdlcEnvironment}.example.com'
            - CertificateCN: !FindInMap [ Global, ALB, CertificateCN ]
              SdlcEnvironment: !FindInMap [ SdlcKey, !Ref SdlcKey, SdlcEnvironment ]
          - !Sub
            - '${CertificateCN}.example.com'
            - CertificateCN: !FindInMap [ Global, ALB, CertificateCN ]
        - !Sub
          - '${CertificateCN}-${SdlcEnvironment}.example.com'
          - CertificateCN: !FindInMap [ Global, ALB, CertificateCN ]
            SdlcEnvironment: !FindInMap [ SdlcKey, !Ref SdlcKey, SdlcEnvironment ]
```

**Comment by kddejong** [comment]
The problem is related to you not producing an array in the else path of the condition.

Change
```yaml
        - !Sub
          - '${CertificateCN}-${SdlcEnvironment}.example.com'
          - CertificateCN: !FindInMap [ Global, ALB, CertificateCN ]
            SdlcEnvironment: !FindInMap [ SdlcKey, !Ref SdlcKey, SdlcEnvironment ]
```
to 
```yaml
        - - !Sub
            - '${CertificateCN}-${SdlcEnvironment}.example.com'
            - CertificateCN: !FindInMap [ Global, ALB, CertificateCN ]
              SdlcEnvironment: !FindInMap [ SdlcKey, !Ref SdlcKey, SdlcEnvironment ]
```

**Comment by JHBoricua** [comment]
hi @kddejong , 

Changing the 'else' block as you suggest didn't get rid of the error.  I might add that this is already a deployed resource in Cloudformation, the template is valid. We simply started updating cfn-lint from v0 to v1 on some of our tooling and started getting linting errors like this one and trying to understand better why that is.

**Comment by JHBoricua** [comment]
My mistake, adding the `- -` in the else  path did get rid of the lint error, my first attempt must have had a typo somewhere. I assume v1 is more strict when it comes to this as the template was deemed valid on cfn-lint v0 and did deploy correctly in AWS?

**Comment by kddejong** [comment]
@JHBoricua v1 is more thorough for sure. What is interesting is that it was deploying correctly when the environment wasn't production.  The new schemas we are using allow for multiple types or variations so it would be possible to have this property accept an array and a string but that's not what they have defined.  

```json
"SubjectAlternativeNames": {
   "items": {
    "type": "string"
   },
   "type": "array",
   "uniqueItems": true
  }
```

Worth noting that [early validation](https://aws.amazon.com/about-aws/whats-new/2024/03/aws-cloudformation-new-validation-checks-stack-operations/) will work after the parameters and most functions are resolved. So any chance that its never possible that `AccountIsCorrect` is True and `IsProd` is False?

**Comment by kddejong** [comment]
When deploying the template with `AccountIsCorrect` is True and `IsProd` is False I get the following error `Value of property SubjectAlternativeNames must be of type List of String`

**Comment by JHBoricua** [comment]
> So any chance that its never possible that `AccountIsCorrect` is True and `IsProd` is False?

Not when deploying resources to our non-prod environments. I assume you are seeing that error with the original version I posted, not the updated one you suggested?

**Comment by JHBoricua** [comment]
Closing this as it is behaving as expected.
